rasa/core/nlg/generator.py

```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
             f"`NaturalLanguageGenerator` instance. Error: {e}"
         )
+
+
+class ResponseVariationFilter:
+    """Filters response variations based on the channel, action and condition."""
+
+    def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
+        self.responses = responses
+
+    @staticmethod
+    def _matches_filled_slots(
+        filled_slots: Dict[Text, Any], response: Dict[Text, Any]
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get(RESPONSE_CONDITION, [])
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            filled_slots_value = filled_slots.get(name)
+            if isinstance(filled_slots_value, str) and isinstance(value, str):
+                if filled_slots_value.casefold() != value.casefold():
+                    return False
+            elif filled_slots_value != value:
+                return False
+
+        return True
+
+    def responses_for_utter_action(
+        self,
+        utter_action: Text,
+        output_channel: Text,
+        filled_slots: Dict[Text, Any],
+    ) -> List[Dict[Text, Any]]:
+        """Returns array of responses that fit the channel, action and condition."""
+        default_responses = list(
+            filter(
+                lambda x: (x.get(RESPONSE_CONDITION) is None),
+                self.responses[utter_action],
+            )
+        )
+        conditional_responses = list(
+            filter(
+                lambda x: (
+                    x.get(RESPONSE_CONDITION)
+                    and self._matches_filled_slots(
+                        filled_slots=filled_slots, response=x
+                    )
+                ),
+                self.responses[utter_action],
+            )
+        )
+
+        conditional_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), conditional_responses)
+        )
+        conditional_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), conditional_responses)
+        )
+        default_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), default_responses)
+        )
+        default_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), default_responses)
+        )
+
+        if conditional_channel:
+            return conditional_channel
+
+        if default_channel:
+            return default_channel
+
+        if conditional_no_channel:
+            return conditional_no_channel
+
+        return default_no_channel
+
+    def get_response_variation_id(
+        self,
+        utter_action: Text,
+        tracker: DialogueStateTracker,
+        output_channel: Text,
+    ) -> Optional[Text]:
+        """Returns the first matched response variation ID that fits the channel, action and condition."""  # noqa: E501
```

Comment:
I'm not sure why `noqa: E501` was required here. Can't we just split the docstring in multiple lines?

Reply:
I can try but I remember the linter complains if it doesn't encounter a `.` period char in the first line.